src/pybind/mgr/orchestrator/module.py

```diff
@@ -667,9 +667,15 @@ def _apply_osd(self,
         if inbuf:
             if unmanaged is not None:
                 return HandleCommandResult(-errno.EINVAL, stderr=usage)
+
             try:
-                drivegroups = yaml.safe_load_all(inbuf)
+                drivegroups = [_dg for _dg in yaml.safe_load_all(inbuf)]
+            except yaml.scanner.ScannerError as e:
```

Comment:
hm, we have this already a few lines below:

```
except ValueError as e:
    msg = 'Failed to read JSON/YAML input: {}'.format(str(e)) + usage
    return HandleCommandResult(-errno.EINVAL, stderr=msg)
```
What happened here?

Reply:
:) Two things. Forstly, the yaml_safe_load returns a generator, so if you don't validate the complete doc up front, you can end up with a partial deployment - by creating a list the yaml is parsed for all documents at the start to prevent this.

Secondly, if the doc does contain a yaml error (I tested by simply not having a space between a key and value), the ValueError doesn't catch it and the user is treated to the traceback - which is why the yaml.scanner.ScannerError exception has been added.
@sebastian-philipp ^^

Reply:
ah, then you can simply change:

```python
except ValueError as e:
```
to:
```python
except (ValueError, yaml.scanner.ScannerError) as e:
```

below. 

No need to have two exception handlers handling the same exception

Reply:
Not really. By creating the list from the generator and  catching the exception we validate the contents of the yaml up front - before any action is taken. What I think we should avoid is bad yaml where some 'docs' are actioned, and then other fails and the admin has to start hacking bits out of the yaml for resubmit. My intent here is to validate up front - and only act on valid input!

Reply:
ok. can you then fix/remove the second exception handler?

Reply:
done in latest commit.